oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/reactive/function/client/ServletOAuth2AuthorizedClientExchangeFilterFunction.java

```diff
@@ -364,103 +367,59 @@ private void populateDefaultOAuth2AuthorizedClient(Map<String, Object> attrs) {
 					.loadAuthorizedClient(clientRegistrationId, authentication,
 							request);
 			if (authorizedClient == null) {
-				authorizedClient = getAuthorizedClient(clientRegistrationId, attrs);
+				ClientRegistration clientRegistration = this.clientRegistrationRepository.findByRegistrationId(
+						clientRegistrationId);
+				if (clientRegistration == null) {
+					throw new IllegalArgumentException("Could not find ClientRegistration with id " + clientRegistrationId);
+				}
+				authorizedClient = authorizeClient(clientRegistration, attrs);
 			}
 			oauth2AuthorizedClient(authorizedClient).accept(attrs);
 		}
 	}
 
-	private OAuth2AuthorizedClient getAuthorizedClient(String clientRegistrationId, Map<String, Object> attrs) {
-		ClientRegistration clientRegistration = this.clientRegistrationRepository.findByRegistrationId(clientRegistrationId);
-		if (clientRegistration == null) {
-			throw new IllegalArgumentException("Could not find ClientRegistration with id " + clientRegistrationId);
-		}
-		if (isClientCredentialsGrantType(clientRegistration)) {
-			return authorizeWithClientCredentials(clientRegistration, attrs);
+	private Mono<OAuth2AuthorizedClient> authorizedClient(Map<String, Object> attrs) {
+		OAuth2AuthorizedClient authorizedClient = getOAuth2AuthorizedClient(attrs);
+		if (reauthorizeRequired(authorizedClient)) {
```

Comment:
Would this logic of determining whether or not to refresh be better suited inside the authorization provider? Doing so would relieve zuul filters, gateway filters and the like from having to replicate it.